### DISCUSSION TITLE
How does emscripten_console_xxx work?

### DISCUSSION THREAD

**Discussion by msqr1** [emscripten-core/emscripten#21783] [Q&A] [answered]
I follow the function hierarchy for it, and it just calls printf_core. How does it skip the fs subsystem? It calls the function printf calls?

**Comment by brendandahl** [comment]
The non formatting versions directly call `console.log` etc [here](https://github.com/emscripten-core/emscripten/blob/ffb52c98fe2ccc07081c3bded2b676ec6f8a6fbe/src/library.js#L2970-L2989).

**Comment by msqr1** [comment]
What about the format version? If it calls printf then what advantage does it offer over regular printf

**Comment by brendandahl** [comment]
The format versions use[ special compiler format attributes](https://releases.llvm.org/5.0.0/tools/clang/docs/AttributeReference.html#format-gnu-format) that don't actually go through printf.

**Comment by sbc100** [comment, marked as answer]
The formatting is done to a string, rather than to stdout (using sprintf).   The result of the formatting is that sent directly to JS: https://github.com/emscripten-core/emscripten/blob/d9aed286296e04b80f05b6f827d7434d1a3e115f/system/lib/libc/emscripten_console.c#L13-L28

If you want even more low level logging (that can also bypass the console object and write directly to stdout/stderr on node) you can also use `emscripten_out` / `emscripten_err` / `emscripten_dbg` and the `f` versions of those 3.    Those function correspond directly the emscripten `err` / `out` / `dbg` helpers in JavaScript.

**Comment by msqr1** [comment]
So the format version just format on a string, and send that string to the normal version

**Reply by sbc100** [reply]
Exactly!

**Reply by msqr1** [reply]
I get it, so then now how do we simulate normal printf? In other words, how to simulate a stdout environment on the browser? Ik that on nodejs stdout is the processes stdout, but wouldn't that just be the browser console, therefore it can be done the same way?

**Reply by sbc100** [reply]
You can just use the actual `printf` function.. that writes to stdout.   The stdout of emscripten will go to the console by default by you can override it by providing `Module['print']` yourself and then you can do whatever you want with it.    See https://emscripten.org/docs/api_reference/module.html#module-object for how to provide Module overrides like this.

**Reply by sbc100** [reply]
If just want the output to go the DevTools console then you don't need to do anything, and printf should just work.

**Reply by msqr1** [reply]
So stdout is an abstract object, it will be console on the browser, and the system stdout in node,... emscripten_console_xxx is only for browser, and it bypass the abstract stdout.

**Reply by sbc100** [reply]
Exactly.   

Although nodejs also supports `console` to the `emscripten_console_xx` APIs works fine there too.

**Reply by msqr1** [reply]
Oh great! Thanks very much!
